src/wallet/wallet.cpp

```diff
@@ -1683,14 +1707,20 @@ CBlockIndex* CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, CBlock
             if (pindex == pindexStop) {
                 break;
             }
-            pindex = chainActive.Next(pindex);
+            {
+                LOCK(cs_main);
+                pindex = chainActive.Next(pindex);
```

Comment:
I think there is still a bug here where scan will return success (`nullptr`) if it gets aborted due to a reorg.

Maybe this line:

```c++
pindex = chainActive.Next(pindex);
```

should be changed to something like:

```c++
if (CBlockIndex* new_index = chainActive.Next(pindex)) {
    pindex = new_index;
} else {
    /* aborting, reorg */
    ret = pindex;
    break;
}
```

Another alternative could be to resume instead of aborting on reorgs.

Reply:
> should be changed to something like:

I think my suggestion above won't work because it will return an error when pindex points to the the tip. Maybe simplest fix would be to restructure the code a little to just acquire cs_main once after ReadBlockFromDisk instead of acquiring, releasing, then immediately re-acquiring. That way there is would be no chance of reorg between the `chainActive.Contains` check above and this `chainActive.Next` call.

Reply:
> I think there is still a bug here where scan will return success

Sorry for the churn, but I no longer think there's a bug here. Opened #12275 to try to clarify the return value.